common/src/attestations/signers.rs

```diff
@@ -1,32 +1,39 @@
 // Copyright 2023-, Edge & Node, GraphOps, and Semiotic Labs.
 // SPDX-License-Identifier: Apache-2.0
 
-use eventuals::{join, Eventual, EventualExt};
 use std::collections::HashMap;
 use std::sync::Arc;
 use thegraph_core::{Address, ChainId};
-use tokio::sync::Mutex;
+use tokio::sync::{watch, watch::Receiver, Mutex};
 use tracing::warn;
 
 use crate::prelude::{Allocation, AttestationSigner};
 
 /// An always up-to-date list of attestation signers, one for each of the indexer's allocations.
 pub fn attestation_signers(
-    indexer_allocations: Eventual<HashMap<Address, Allocation>>,
+    mut indexer_allocations: Receiver<HashMap<Address, Allocation>>,
     indexer_mnemonic: String,
     chain_id: ChainId,
-    dispute_manager: Eventual<Address>,
-) -> Eventual<HashMap<Address, AttestationSigner>> {
+    mut dispute_manager: Receiver<Address>,
+) -> Receiver<HashMap<Address, AttestationSigner>> {
     let attestation_signers_map: &'static Mutex<HashMap<Address, AttestationSigner>> =
         Box::leak(Box::new(Mutex::new(HashMap::new())));
-
     let indexer_mnemonic = Arc::new(indexer_mnemonic);
 
-    // Whenever the indexer's active or recently closed allocations change, make sure
-    // we have attestation signers for all of them
-    join((indexer_allocations, dispute_manager)).map(move |(allocations, dispute_manager)| {
-        let indexer_mnemonic = indexer_mnemonic.clone();
-        async move {
+    let (tx, rx) = watch::channel(HashMap::new());
+
+    tokio::spawn(async move {
+        loop {
+            tokio::select! {
+                Ok(_) = indexer_allocations.changed() => {},
+                Ok(_) = dispute_manager.changed() => {},
+                else => break,
+            }
+
```

Comment:
One suggestion to keep the dispute-manager isolated from `Eventual`s is to use:
```rust
let subscriber = eventual.subscribe();

select! {
    value = subscriber.read() = {}
    ...
```

This way you can rebuild the `join` without needing two eventuals.

I might be wrong on this one, let me know how this goes.

Reply:
was outside yesterday, thanks for the review BTW, I'll push the update today itself ✌️

Reply:
Sorry for the delay[🥲], subscriber didn't work as expected, so had to use pipe to listen to the changes, which in turn will update a new watch::channel for indexer_allocations, hope this will be good as eventually we will be migrating all to tokio  watch

Reply:
BTW great to know eventuals was build at edge&node itself 🔥

Reply:
eventuals::join would only work if both variables are updated right?, so if you'd like we could modify it to work, when either one is updated

Reply:
yes, exactly

Reply:
actually, it updates when either one updates as long as both are defined. It only joins when both are Some

Reply:
Ok, I might have made some mistakes in testing the function, I'll send the updated pr asap